File: android_app/app/src/main/java/com/oceansentinels/app/presentation/viewmodel/ThemeViewModel.kt
package com.oceansentinels.app.presentation.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.oceansentinels.app.data.local.preferences.PreferencesManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * ViewModel to manage app-wide theme (dark mode) and offline mode state.
 * Reads from and writes to PreferencesManager (DataStore).
 */
@HiltViewModel
class ThemeViewModel @Inject constructor(
    private val preferencesManager: PreferencesManager
) : ViewModel() {

    val isDarkMode: StateFlow<Boolean> = preferencesManager.isDarkMode
        .stateIn(viewModelScope, SharingStarted.Eagerly, false)

    val isOfflineMode: StateFlow<Boolean> = preferencesManager.isOfflineMode
        .stateIn(viewModelScope, SharingStarted.Eagerly, false)

    fun toggleDarkMode(enabled: Boolean) {
        viewModelScope.launch {
            preferencesManager.setDarkMode(enabled)
        }
    }

    fun toggleOfflineMode(enabled: Boolean) {
        viewModelScope.launch {
            preferencesManager.setOfflineMode(enabled)
        }
    }
}

File: IOS/OceanSentinels/Presentation/ViewModels/ThemeViewModel.swift
import Foundation
import Observation

// MARK: - ThemeViewModel

/// Manages app-wide theme (dark mode) and offline mode preferences.
///
/// Reads/writes through `PreferencesManager` which persists to UserDefaults.
@Observable
@MainActor
final class ThemeViewModel {

    // MARK: - State

    var isDarkMode: Bool
    var isOfflineMode: Bool

    // MARK: - Dependencies

    private let preferencesManager: PreferencesManager

    // MARK: - Init

    init(preferencesManager: PreferencesManager = .shared) {
        self.preferencesManager = preferencesManager
        self.isDarkMode = preferencesManager.isDarkMode
        self.isOfflineMode = preferencesManager.isOfflineMode
    }

    // MARK: - Actions

    func toggleDarkMode(_ enabled: Bool) {
        isDarkMode = enabled
        preferencesManager.isDarkMode = enabled
    }

    func toggleOfflineMode(_ enabled: Bool) {
        isOfflineMode = enabled
        preferencesManager.isOfflineMode = enabled
    }
}
